### PR TITLE
Template unroll for RCCL kernels

### PR DESCRIPTION
**Pull request by wenkaidu**
## Details
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** _"Internal", or link to GitHub issue (if applicable)._

**What were the changes?**  
Template unroll for RCCL kernels

**Why were the changes made?**  
Allow RCCL kernels to be built with different unrolls

**How was the outcome achieved?**  
Template unroll for RCCL kernels

**Additional Documentation:**  
_What else should the reviewer know?_

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

File: src/graph/xml.cc [hunk 1]
```diff
@@ -651,11 +651,9 @@ ncclResult_t ncclTopoGetXmlFromGpu(struct ncclXmlNode* pciNode, uint32_t rocmDev
 
   NCCLCHECK(xmlGetAttrIndex(gpuNode, "sm", &index));
   if (index == -1) {
-    int cudaMajor, cudaMinor;
     cudaDeviceProp devProp;
     CUDACHECK(cudaGetDeviceProperties(&devProp, 0));
-    cudaMajor = devProp.major; cudaMinor = devProp.minor;
-    NCCLCHECK(xmlSetAttrInt(gpuNode, "sm", cudaMajor*10+cudaMinor));
+    NCCLCHECK(xmlSetAttrInt(gpuNode, "sm", devProp.multiProcessorCount));
```
Comment (nusislam) [inline src/graph/xml.cc:656]: Why is this change needed?

File: src/graph/topo.cc [hunk 2]
```diff
@@ -379,6 +379,8 @@ ncclResult_t ncclTopoAddGpu(struct ncclXmlNode* xmlGpu, struct ncclTopoSystem* s
 #if defined(__HIP_PLATFORM_AMD__) || defined(__HCC__) || defined(__HIPCC__)
   // There is no direct mapping between CUDA SM to HIP GFX. Use SM60 as compatibility level.
   gpu->gpu.cudaCompCap = 60;
+  // Repurpose previously unused "sm" as CU counts
+  NCCLCHECK(xmlGetAttrInt(xmlGpu, "sm", &gpu->gpu.cu));
```
Comment (nusislam) [inline src/graph/topo.cc:383]: I wonder whether any of these changes will have impact for the CPX mode.

File: src/graph/connect.cc [hunk 3]
```diff
@@ -631,7 +631,8 @@ ncclResult_t ncclTopoPostset(struct ncclComm* comm, int* firstRanks, int* treePa
   // Define channels for gfx94 GPU architectures
   if (IsArchMatch(comm->topo->nodes[GPU].nodes[0].gpu.gcn, "gfx94")) {
     // Only use full MAXCHANNELS for gfx94x
-    maxChannels = MAXCHANNELS;
+    // Use max 80 channels in case of 80 CUs
+    maxChannels = comm->topo->nodes[GPU].nodes[0].gpu.cu == 80 ? 80 : MAXCHANNELS;
```
Comment (corey-derochie-amd) [inline src/graph/connect.cc]: Is 80 special, or should this be something like
```suggestion
    maxChannels = min(comm->topo->nodes[GPU].nodes[0].gpu.cu, MAXCHANNELS);
```

File: src/graph/xml.cc (hunk 1 above)
Comment (corey-derochie-amd) [inline src/graph/xml.cc:656]: Why reuse "sm" for something different?

**Review by corey-derochie-amd — APPROVED** [review]
LGTM + some questions

File: src/graph/connect.cc (hunk 3 above)
Comment (wenkaidu) [reply src/graph/connect.cc]: since this PR reuses "sm" field, it contains some random numbers like "90", "94", "98"... in older XMLs.

File: src/graph/xml.cc (hunk 1 above)
Comment (wenkaidu) [reply src/graph/xml.cc:656]: CU count needs to be put into xml. I didn't want to create new xml just for one field

File: src/graph/topo.cc (hunk 2 above)
Comment (wenkaidu) [reply src/graph/topo.cc:383]: If CPX implementation needs to know CU count, this will help

File: src/graph/xml.cc (hunk 1 above)
Comment (wenkaidu) [reply src/graph/xml.cc:656]: "sm" is not really valid for AMD GPU

File: src/graph/connect.cc (hunk 3 above)
Comment (wenkaidu) [reply src/graph/connect.cc]: alternatively, we need to fix all existing xmls with CU count

**Review by nusislam — APPROVED** [review]
